### PR TITLE
URL Shortner feature initial version

### PR DESCRIPTION
**Pull request by harshith-venkatesh**
I am currently confused with using Regex or using valid-url package
https://www.npmjs.com/package/valid-url
Open To Suggestion
[X] Tests to be added

### CODE REVIEW

File: models/urls.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Add url
+ *
+ * @param urlData {Object}: url data object to be stored in DB
+ * @return {Promise<{urlData: Object}>}
+ */
+const addUrl = async (urlData) => {
+  try {
+    const { id } = await urlsModel.add(urlData)
+    return { id, longUrl: urlData.longUrl, shortUrl: urlData.shortUrl }
```
Comment (swarajpure) [inline models/urls.js:16]: we can destructure `urlData` already

File: models/urls.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    throw err
+  }
+}
+
+const fetchUrl = async (shortUrl) => {
+  try {
+    const snapshot = await urlsModel.where('shortUrl', '==', shortUrl).get()
+    const [urlData] = snapshot.docs
+    if (urlData) {
+      return { id: urlData.id, longUrl: urlData.longUrl, shortUrl: urlData.shortUrl }
```
Comment (swarajpure) [inline models/urls.js:48]: re: destructure

**Review by swarajpure — CHANGES_REQUESTED** [review]
Some destructuring can be done
Please check 🙂

File: controllers/urlShortner.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+const { nanoid } = require('nanoid')
+const urlQuery = require('../models/urls')
+
+const ERROR_MESSAGE = 'Error getting urls'
+const NOTVALIDURL = 'Please attach valid URL'
+const NOTVALIDSHORTURL = 'Please attach valid ID'
+const validUrlRegex = /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:[/?#]\S*)?$/i
```
Comment (harshith-venkatesh) [inline controllers/urlShortner.js:7]: 7:23  warning  Unsafe Regular Expression  security/detect-unsafe-regex
Currently receiving this warning 
I tried to resolve this issue but have not found the one which is resolving this warning
An alternative would be to install valid-url package which I did found would be overload which could be resolved with above statement

Open to suggestions :)

**Review by harshith-venkatesh — COMMENTED** [review]
@ankurnarkhede @sumitd94 @swarajpure

File: controllers/urlShortner.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+const { nanoid } = require('nanoid')
+const urlQuery = require('../models/urls')
+
+const ERROR_MESSAGE = 'Error getting urls'
+const NOTVALIDURL = 'Please attach valid URL'
```
Comment (Neha) [inline controllers/urlShortner.js:5]: The words should be separated by the `underscore`

Comment (Neha) [inline controllers/urlShortner.js:5]: The words should be separated by the `underscore`

File: controllers/urlShortner.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    const { longUrl } = req.body
+    const fetchAvailableUrl = await urlQuery.checkUrlExists(longUrl)
+    if (fetchAvailableUrl) {
+      return res.json({
+        message: 'Url returned successfully',
+        urlData: fetchAvailableUrl
+      })
+    }
+    if (validUrlRegex.test(longUrl)) {
+      const shortUrl = 'RDS' + nanoid(6)
```
Comment (Neha) [inline controllers/urlShortner.js:45]: `RDS` should be moved to constant.

File: controllers/urlShortner.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      const urlData = await urlQuery.addUrl({
+        shortUrl: shortUrl,
+        longUrl: longUrl
+      })
+      return res.json({
+        message: 'Url created successfully',
```
Comment (Neha) [inline controllers/urlShortner.js:51]: URL created...should be moved to Constant

File: models/urls.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Fetch Url
+ *
+ * @param longUrl {string}: url string to fetch urlData
+ * @return {Promise<{urlData|Object}>}
+ */
+const checkUrlExists = async (longUrl) => {
+  try {
+    const snapshot = await urlsModel.where('longUrl', '==', longUrl).get()
```
Comment (Neha) [inline models/urls.js:32]: `snapshot` of what? the variable name should be descriptive.

File: controllers/urlShortner.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ *
+ * @param req {String} - Exxpress request object
+ * @param res {Object} - Express response object
+ */
+
+const getLongUrl = async (req, res) => {
+  try {
+    const { shortUrl } = req.params
+    const fetchUrlData = await urlQuery.fetchUrl(shortUrl)
```
Comment (Neha) [inline controllers/urlShortner.js:17]: Question: What is the significance of the `data` in `fetchUrlData` here?

File: controllers/urlShortner.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    if (fetchAvailableUrl) {
+      return res.json({
+        message: 'Url returned successfully',
+        urlData: fetchAvailableUrl
+      })
+    }
+    if (validUrlRegex.test(longUrl)) {
+      const shortUrl = 'RDS' + nanoid(6)
+      const urlData = await urlQuery.addUrl({
+        shortUrl: shortUrl,
```
Comment (Neha) [inline controllers/urlShortner.js:47]: As the key and value are the same we can remove the value and keep it  (https://javascript.info/object#property-value-shorthand)

{ 
    shortUrl,
  longUrl
}

**Comment by prakashchoudhary07** [comment]
Closing this as we are having a separate backend for url-shortener in go #tiny-site-backend

**Pull request closed without merging**